service/migration_manager.cc

```diff
@@ -1145,27 +1145,38 @@ future<schema_ptr> migration_manager::get_schema_for_write(table_schema_version
     }
 
     auto s = local_schema_registry().get_or_null(v);
-
-    if (s && s->is_synced()) {
-        co_return s;
-    }
-
     // `_enable_schema_pulls` may change concurrently with this function (but only from `true` to `false`).
     bool use_raft = !_enable_schema_pulls;
-
-    if (use_raft) {
-        // Schema is synchronized through Raft, so perform a group 0 read barrier.
-        // Batch the barriers so we don't invoke them redundantly.
+    if ((!s || !s->is_synced()) && use_raft) {
+        // batch group0 raft barriers
         co_await (as ? _group0_barrier.trigger(*as) : _group0_barrier.trigger());
     }
-
-    s = co_await get_schema_definition(v, dst, ms, _storage_proxy);
-
-    if (use_raft) {
-        // If Raft is used the schema is synced already (through barrier above), mark it as such.
-        co_await s->registry_entry()->maybe_sync([] { return make_ready_future<>(); });
-    } else {
-        co_await maybe_sync(s, dst);
+    if (!s) {
+        s = co_await get_schema_definition(v, dst, ms, _storage_proxy);
+    }
+    if (!s->is_synced()) {
+        if (use_raft) {
+            // Mark entry as synced we have already aquired the barrier
+            co_await s->registry_entry()->maybe_sync([] { return make_ready_future<>(); });
+        } else {
+            co_await maybe_sync(s, dst);
+        }
+    }
+    // here s is guaranteed to be valid and synced
+    if (s->is_view() && !s->view_info()->base_info()) {
```

Comment:
How does this fix prevent the problem after deactivation->activation?

Reply:
The code makes sure that we either except out of `get_schema_for_write` or return an activated schema with
the base info.

Reply:
So this fix assumes that every activation goes through get_schema_for_write().

Reply:
Not exactly.
The context of when the activation happened is not that important.
This fix "knows" however, that callers to get_schema_for_write/get_schema_for_read are going to create global pointers from the obtained schema and that there is a chance that the schema deactivated (and maybe even reactivated) in the past. So all that this fix boils down to is to make sure that schemas that are returned from the aforementioned functions can be globalized.

Reply:
The cover letter first sentence is:
> This miniset addresses two potential conversions to global_schema_ptr of incomplete materialized views schemas.

So it is a specific fix, not a fix all solution which will probably be complicated but I agree we should eventually have.

Reply:
Yes, I approved the fix because it fixes some common scenarios and doesn't make things worse. But I think we still have holes which may bite us and we should come with a more systemic approach for handling view schemas.